### PR TITLE
Add service discovery for OvhCloud

### PR DESCRIPTION
**Pull request by Mama59**
This PR adds a new service disocvery for [Ovhcloud Bare Metal Cloud](https://www.ovhcloud.com/en/bare-metal/) based on the work that has been done for Ionos and Scaleway.
The service discovery allows users to create scrape targets based on servers tied to an OVHcloud account.

Thank you for your time on this PR.
Answered https://github.com/prometheus/prometheus/pull/10775#issuecomment-1144080466
Signed-off-by: Marine Bal [marine.bal@ovhcloud.com](mailto:marine.bal@ovhcloud.com)

### CODE REVIEW

File: discovery/ovhcloud/ovhcloud.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+
+	validate := validator.New()
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:100]: Why is this needed?

File: discovery/ovhcloud/ovhcloud.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+func (c SDConfig) NewDiscoverer(options discovery.DiscovererOptions) (discovery.Discoverer, error) {
+	return NewDiscovery(&c, options.Logger)
+}
+
+func init() {
+	discovery.RegisterConfig(&SDConfig{})
+}
+
+// ParseIPList Parse ip list to store on IPV4 and IPV6 on IPs type
+func ParseIPList(ipList []string) (*IPs, error) {
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:135]: Why is this needed? If the API does not make the difference between ipv4 and ipv6, I don't think we should do it downstream.

File: discovery/ovhcloud/ovhcloud.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	}
+
+	client, err := c.CreateClient()
+	if err != nil {
+		return err
+	}
+
+	if !c.NoAuthTest {
+		var authDetails AuthDetails
+		fmt.Print("Test auth/details\n")
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:113]: This is debug code.

File: discovery/ovhcloud/ovhcloud.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+// AuthDetails partial
+type AuthDetails struct {
+	User string `json:"user"`
+}
+
+// SDConfig sd config
+type SDConfig struct {
+	Endpoint          string         `yaml:"endpoint" validate:"required"`
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:72]: We do not use the validate library in Prometheus.

File: discovery/ovhcloud/ovhcloud.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+	ApplicationKey    string         `yaml:"application_key" validate:"required"`
+	ApplicationSecret string         `yaml:"application_secret" validate:"required"`
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:74]: this should be a config.Secret, not a string

File: discovery/ovhcloud/dedicated_server.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	client, err := (d.config).CreateClient()
+	if err != nil {
+		return nil, err
+	}
+	var dedicatedServerDetailedList []DedicatedServer
+	dedicatedServerList, err := getDedicatedServerList(client)
+	if err != nil {
+		return nil, err
+	}
+	for _, dedicatedServerName := range dedicatedServerList {
```
Comment (roidelapluie) [inline discovery/ovhcloud/dedicated_server.go:113]: This will generate a lot of API calls, is it needed to get the details for each individual hosts?

File: discovery/ovhcloud/ovhcloud_test.go [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+package ovhcloud
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"gopkg.in/h2non/gock.v1"
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud_test.go:23]: I wonder if it is worth it to introduce this dependency in Prometheus.

**Review by roidelapluie — COMMENTED** [review]
Thanks for this!!

I have done a first high level review. In general, I like to limit the introduction of new dependencies like validator and gock, and probably also void auto generating labels from structs. We should also look into reducing the number of API calls.

**Comment by roidelapluie** [comment]
Also, you can just push code to your branch without recreating a new pull request each time. Thanks!

File: discovery/ovhcloud/ovhcloud.go (hunk 2 above)
Comment (Arnoways) [reply discovery/ovhcloud/ovhcloud.go:135]: We use it to add ipv4 and ipv6 as meta in discovered labels so users can pick the one they want to use for scrapping.

File: discovery/ovhcloud/ovhcloud.go [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+	ConsumerKey       string         `yaml:"consumer_key" validate:"required"`
+	RefreshInterval   model.Duration `yaml:"refresh_interval" validate:"required"`
+	Service           string         `yaml:"service" validate:"required,oneof=dedicated_server vps"`
+	NoAuthTest        bool           `yaml:"no_auth_test"`
+}
+
+// IPs struct to store IPV4 and IPV6
+type IPs struct {
+	IPV4 string `json:"ipv4" label:"ipv4"`
```
[the remark this answers is not in the document]
Comment (Arnoways) [reply discovery/ovhcloud/ovhcloud.go:83]: The goal is to have the ips addresses as meta, such as:
image

File: discovery/ovhcloud/dedicated_server.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const dedicatedServerAPIPath = "/dedicated/server"
+
+// DedicatedServer struct
+type DedicatedServer struct {
```
[the remark this answers is not in the document]
Comment (Arnoways) [reply discovery/ovhcloud/dedicated_server.go:34]: No, it's a simple wrapper over the API.

**Comment by roidelapluie** [comment]
In general, labels should be as close as possible to the fields the API
offers.

On 03 Jun 04:53, Arnoways wrote:
> @Arnoways commented on this pull request.
> 
> 
> 
> > +// CreateClient get client
> +func (c SDConfig) CreateClient() (*ovh.Client, error) {
> +	return ovh.NewClient(c.Endpoint, c.ApplicationKey, c.ApplicationSecret, c.ConsumerKey)
> +}
> +
> +// NewDiscoverer new discoverer
> +func (c SDConfig) NewDiscoverer(options discovery.DiscovererOptions) (discovery.Discoverer, error) {
> +	return NewDiscovery(&c, options.Logger)
> +}
> +
> +func init() {
> +	discovery.RegisterConfig(&SDConfig{})
> +}
> +
> +// ParseIPList Parse ip list to store on IPV4 and IPV6 on IPs type
> +func ParseIPList(ipList []string) (*IPs, error) {
> 
> We use it to add ipv4 and ipv6 as meta in discovered labels so users can pick the one they want to use for scrapping.
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/prometheus/prometheus/pull/10802#discussion_r888873453
> You are receiving this because you commented.
> 
> Message ID: ***@***.***>

-- 
Julien Pivotto
O11y - https://o11y.eu/

File: discovery/ovhcloud/dedicated_server.go (hunk 6 above)
Comment (Arnoways) [reply discovery/ovhcloud/dedicated_server.go:113]: We have to in order to at least get the ip address of the servers. Additionally, it fetches the extra details that will be exposed as discovered _meta_ labels. 
Though we could remove the call to the `ips` route, preventing users from choosing between ipv4 or ipv6 but reducing the amount of API calls.

File: discovery/ovhcloud/ovhcloud.go (hunk 1 above)
Comment (Mama59) [reply discovery/ovhcloud/ovhcloud.go:100]: Is to check validate tag on struct (required, oneof [validator](https://github.com/go-playground/validator))
https://github.com/prometheus/prometheus/pull/10802/commits/12f18c2772a5a9a2499877572c76bf77628c49d8#diff-643d27d894488ad9f1633a2542dd9fe466432bf97880359a5eaab2b9ae8ea4e7R78

Like this 
https://github.com/prometheus/prometheus/pull/10802/commits/4767a7e265c845efaf3fe8fcdeccedaf3062052a#diff-643d27d894488ad9f1633a2542dd9fe466432bf97880359a5eaab2b9ae8ea4e7R77
To do this kind of test with a validate tag
https://github.com/prometheus/prometheus/blob/main/discovery/scaleway/scaleway.go#L65

File: discovery/ovhcloud/ovhcloud.go [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+// metaLabelPrefix is the meta prefix used for all meta labels.
+// in this discovery.
+const (
+	metaLabelPrefix = model.MetaLabelPrefix + "ovhcloud_"
+)
+
+func addFieldsOnLabels(fields []*structs.Field, labels model.LabelSet, prefix string) {
```
[the remark this answers is not in the document]
Comment (Mama59) [reply discovery/ovhcloud/ovhcloud.go:42]: Is to do this 
https://github.com/prometheus/prometheus/blob/main/discovery/scaleway/instance.go#L38
without copying every field manually

File: discovery/ovhcloud/ovhcloud.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Mama59) [reply discovery/ovhcloud/ovhcloud.go:83]: It used here https://github.com/prometheus/prometheus/pull/10802/commits/12f18c2772a5a9a2499877572c76bf77628c49d8#diff-643d27d894488ad9f1633a2542dd9fe466432bf97880359a5eaab2b9ae8ea4e7R45

File: discovery/ovhcloud/ovhcloud_test.go (hunk 7 above)
Comment (Arnoways) [reply discovery/ovhcloud/ovhcloud_test.go:23]: We feel like it could be a nice addition for the test suite, for the code we wrote but also for the upcoming features from the others contributors. If you think it is not worth it, we can remove it.

**Comment by Mama59** [comment]
> 



> 

For labels, we took the same api name with json tag by default, we only update it for 2-3 fields because one is on a sub-object and on object and it's duplicate
and on other one we have a typo into the api (fixed with V2 but not for now)

File: discovery/ovhcloud/ovhcloud_test.go (hunk 7 above)
Comment (roidelapluie) [reply discovery/ovhcloud/ovhcloud_test.go:23]: I would like to avoid extra dependencies like this.

Service discovery is also used by third parties as libraries, so it impacts way more than Prometheus.

File: discovery/ovhcloud/dedicated_server.go (hunk 6 above)
Comment (roidelapluie) [reply discovery/ovhcloud/dedicated_server.go:113]: Is it an option to have those extra calls as an option? What info can we get with a single call?

**Comment by Arnoways** [comment]
Hi @roidelapluie I updated the PR trying to remove the extra libraries and tidying the tests. Let me know what you think

File: discovery/ovhcloud/dedicated_server.go (hunk 6 above)
Comment (Arnoways) [reply discovery/ovhcloud/dedicated_server.go:113]: Basically none, the first call only returns a list of available servers, that's why we have to go through the loop to get the information about each of those servers.
What we could make optional is the extra call for each servers to get the complete list of their ip addresses.

File: discovery/ovhcloud/ovhcloud_test.go (hunk 7 above)
Comment (Mama59) [reply discovery/ovhcloud/ovhcloud_test.go]: In my opinion, third parties doesn't import test files so they are not affected by libraries on _test.go files

**Comment by Arnoways** [comment]
Hello @roidelapluie , any chance you could take a new look at the modifications please?

File: discovery/ovhcloud/ovhcloud.go [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+// limitations under the License.
+
+package ovhcloud
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"inet.af/netaddr"
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:21]: inet.af/netaddr is deprecated

File: discovery/ovhcloud/ovhcloud.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+
+	"github.com/go-kit/log"
+	"github.com/grafana/regexp"
+	"github.com/ovh/go-ovh/ovh"
+	"github.com/pkg/errors"
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:26]: please use standard "errors"

File: discovery/ovhcloud/ovhcloud.go [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+// metaLabelPrefix is the meta prefix used for all meta labels.
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:35]: can you remove the `()`?

File: discovery/ovhcloud/ovhcloud.go [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+// in this discovery.
+const (
+	metaLabelPrefix = model.MetaLabelPrefix + "ovhcloud_"
+)
+
+var (
+	ovhCloudApplicationKeyTest    = "TDPKJdwZwAQPwKX2"
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:42]: This belongs in tests, not in the regular file

File: discovery/ovhcloud/ovhcloud.go [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+type refresher interface {
+	refresh(context.Context) ([]*targetgroup.Group, error)
+}
+
+var DefaultSDConfig = SDConfig{
+	Endpoint:        "ovh-eu",
+	RefreshInterval: model.Duration(60 * time.Second),
+}
+
+// SDConfig sd config
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:56]: Comments should be sentences and end with a full stop.

File: discovery/ovhcloud/ovhcloud.go [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+type SDConfig struct {
+	Endpoint          string         `yaml:"endpoint"`
+	ApplicationKey    string         `yaml:"application_key"`
+	ApplicationSecret config.Secret  `yaml:"application_secret"`
+	ConsumerKey       string         `yaml:"consumer_key"`
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:61]: Is this a secret?

File: discovery/ovhcloud/ovhcloud.go [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+	RefreshInterval   model.Duration `yaml:"refresh_interval"`
+	Service           string         `yaml:"service"`
+}
+
+// IPs struct to store IPV4 and IPV6
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:66]: Comments should end with full stop.

File: discovery/ovhcloud/ovhcloud.go [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+type IPs struct {
+	IPV4 string `json:"ipv4"`
+	IPV6 string `json:"ipv6"`
+}
+
+// Name get name
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:72]: Comments should end with full stop and be a sentence. In this case, Name implements the Discoverer interface.

File: discovery/ovhcloud/ovhcloud.go [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+	if c.ApplicationKey == "" {
+		return errors.New("application key can not be empty")
+	}
+	if c.ApplicationSecret == "" {
+		return errors.New("application secret can not be empty")
+	}
+	if c.ConsumerKey == "" {
+		return errors.New("consumer key can not be empty")
+	}
+	switch c.Service {
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:98]: I feel like a if would be more readable here.

File: discovery/ovhcloud/ovhcloud.go [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+	case "dedicated_server", "vps":
+		return nil
+	default:
+		return errors.Errorf("unknown service: %v", c.Service)
+	}
+}
+
+// CreateClient get client
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:106]: Comment should be full sentences with a dot.

This one maybe does not need a comment and does not need to be exposed.

File: discovery/ovhcloud/ovhcloud.go [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+	switch conf.Service {
+	case "vps":
+		return newVpsDiscovery(conf, logger), nil
+	case "dedicated_server":
+		return newDedicatedServerDiscovery(conf, logger), nil
+	}
+	return nil, fmt.Errorf("unknown OVHcloud discovery service '%s'", conf.Service)
+}
+
+// NewDiscovery ovhcloud creates a discovery with refresher
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:160]: Full step and should be sentence.

File: discovery/ovhcloud/ovhcloud_test.go [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+	_, err := ParseIPList([]string{"A.b"})
+	require.ErrorContains(t, err, "could not parse IP addresses from list")
+}
+
+func TestParseZeroIPFailed(t *testing.T) {
+	_, err := ParseIPList([]string{"0.0.0.0"})
+	require.ErrorContains(t, err, "could not parse IP addresses from list")
+}
+
+func TestParseVoidIPFailed(t *testing.T) {
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud_test.go:79]: All those tests could be rewritten as one with a cases struct.

File: discovery/ovhcloud/vps.go [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+	Datacenter          []string `json:"datacenter"`
+	Vcore               int      `json:"vcore"`
+	Version             string   `json:"version"`
+	Name                string   `json:"name"`
+	Disk                int      `json:"disk"`
+	Memory              int      `json:"memory"`
+}
+
+// Vps struct from API
+type Vps struct {
```
Comment (roidelapluie) [inline discovery/ovhcloud/vps.go:45]:
```suggestion
type VPS struct {
```

is it necessary to expose this?

File: docs/configuration/configuration.md [hunk 24]
```diff
@@ -1171,6 +1175,65 @@ tls_config:
   [ <tls_config> ]
 ```
 
+### `<ovhcloud_sd_config>`
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:1178]: Probably missing an empty line here

File: docs/configuration/configuration.md [hunk 25]
```diff
@@ -1171,6 +1175,65 @@ tls_config:
...
+OVHcloud SD configurations allow retrieving scrape targets from OVHcloud's [dedicated servers](https://www.ovhcloud.com/en/bare-metal/) and [VPS](https://www.ovhcloud.com/en/vps/) using
+their [API](https://api.ovh.com/).
+Prometheus will periodically check the REST endpoint and create a target for every discovered server.
+The role will try to use the public IPv4 address as default address, if there's none it will try to use the IPv6 one. This may be changed with relabeling.
+For OVHcloud's [public cloud instances](https://www.ovhcloud.com/en/public-cloud/) you can use the [openstack_sd_config](#openstack_sd_config).
+
+#### VPS
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:1185]: Probably missing an empty line here

File: docs/configuration/configuration.md [hunk 26]
```diff
@@ -1171,6 +1175,65 @@ tls_config:
...
+* `__meta_ovhcloud_vps_monitoringIpBlocks`: the Ip blocks for OVH monitoring servers
+* `__meta_ovhcloud_vps_cluster`: the cluster of the server
+* `__meta_ovhcloud_vps_state`: the state of the server
+* `__meta_ovhcloud_vps_name`: the name of the server
+* `__meta_ovhcloud_vps_netbootMode`: the netbootMode of the server
+* `__meta_ovhcloud_vps_memoryLimit`: the memoryLimit of the server
+* `__meta_ovhcloud_vps_offerType`: the offerType of the server
+* `__meta_ovhcloud_vps_vcore`: the vcore of the server
+
+#### Dedicated servers
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:1208]: Probably missing an empty line here

File: docs/configuration/configuration.md [hunk 27]
```diff
@@ -1171,6 +1175,65 @@ tls_config:
...
+* `__meta_ovhcloud_dedicated_server_reverse`: dedicated server reverse
+* `__meta_ovhcloud_dedicated_server_datacenter`: the dedicated datacenter localisation
+* `__meta_ovhcloud_dedicated_server_name`: the dedicated server name
+
+See below for the configuration options for OVHcloud discovery:
+
+```yaml
+# Access key to use. https://api.ovh.com
+application_key: <string>
+application_secret: <string>
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:1227]:
```suggestion
application_secret: <secret>
```

File: docs/configuration/configuration.md [hunk 28]
```diff
@@ -1171,6 +1175,65 @@ tls_config:
...
+consumer_key: <string>
+# API endpoint. https://github.com/ovh/go-ovh#supported-apis
+endpoint: <string>
+# Service of the targets to retrieve. Must be `vps` or `dedicated_server`.
+service: <string>
+# Refresh interval to re-read the resources list.
+refresh_interval: <duration>
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:1234]: please add the default

File: go.mod [hunk 29]
```diff
@@ -168,6 +170,8 @@ require (
 	go.opentelemetry.io/otel/exporters/otlp/internal/retry v1.9.0 // indirect
 	go.opentelemetry.io/otel/metric v0.31.0 // indirect
 	go.opentelemetry.io/proto/otlp v0.18.0 // indirect
+	go4.org/intern v0.0.0-20211027215823-ae77deb06f29 // indirect
+	go4.org/unsafe/assume-no-moving-gc v0.0.0-20220617031537-928513b29760 // indirect
```
Comment (roidelapluie) [inline go.mod:174]: what is using this?

File: discovery/ovhcloud/dedicated_server.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (roidelapluie) [reply discovery/ovhcloud/dedicated_server.go:34]: I don't think we should make this public

**Review by roidelapluie — COMMENTED** [review]
Sorry, let's get back to this! Next round of comments is here.

File: discovery/ovhcloud/ovhcloud.go (hunk 16 above)
Comment (Arnoways) [reply discovery/ovhcloud/ovhcloud.go:61]: It should probably be one. Thank you!

File: discovery/ovhcloud/ovhcloud.go (hunk 19 above)
Comment (Arnoways) [reply discovery/ovhcloud/ovhcloud.go:98]: We went with a switch for consistency since it appears that's how it is done for the others SDs, see:
https://github.com/prometheus/prometheus/blob/main/discovery/hetzner/hetzner.go#L95
https://github.com/prometheus/prometheus/blob/main/discovery/scaleway/scaleway.go#L61

I don't mind the modification, so it's up to you.

File: discovery/ovhcloud/dedicated_server.go [hunk 30]
```diff
@@ -0,0 +1,150 @@
...
+
+func getDedicatedServerDetails(client *ovh.Client, serverName string) (*dedicatedServer, error) {
+	var dedicatedServerDetails dedicatedServer
+	err := client.Get(fmt.Sprintf("%s/%s", dedicatedServerAPIPath, url.QueryEscape(serverName)), &dedicatedServerDetails)
+	if err != nil {
+		return nil, err
+	}
+
+	var ips []string
+	err = client.Get(fmt.Sprintf("%s/%s/ips", dedicatedServerAPIPath, url.QueryEscape(serverName)), &ips)
```
Comment (roidelapluie) [inline discovery/ovhcloud/dedicated_server.go:79]: Can we use path.Join here?

File: discovery/ovhcloud/dedicated_server.go [hunk 31]
```diff
@@ -0,0 +1,150 @@
...
+)
+
+const (
+	dedicatedServerAPIPath     = "/dedicated/server"
+	dedicatedServerLabelPrefix = metaLabelPrefix + "dedicatedServer_"
+)
+
+type dedicatedServer struct {
+	State           string `json:"state"`
+	IPs             IPs    `json:"ips"`
```
Comment (roidelapluie) [inline discovery/ovhcloud/dedicated_server.go:38]: does this need a json tag since we do not get it from the API? Maybe it does not have to be public either

File: discovery/ovhcloud/dedicated_server.go [hunk 32]
```diff
@@ -0,0 +1,150 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return dedicatedListName, nil
+}
+
+func getDedicatedServerDetails(client *ovh.Client, serverName string) (*dedicatedServer, error) {
+	var dedicatedServerDetails dedicatedServer
+	err := client.Get(fmt.Sprintf("%s/%s", dedicatedServerAPIPath, url.QueryEscape(serverName)), &dedicatedServerDetails)
```
Comment (roidelapluie) [inline discovery/ovhcloud/dedicated_server.go:73]: Can we use path.Join here?

File: discovery/ovhcloud/ovhcloud.go [hunk 33]
```diff
@@ -0,0 +1,163 @@
...
+	Endpoint          string         `yaml:"endpoint"`
+	ApplicationKey    string         `yaml:"application_key"`
+	ApplicationSecret config.Secret  `yaml:"application_secret"`
+	ConsumerKey       config.Secret  `yaml:"consumer_key"`
+	RefreshInterval   model.Duration `yaml:"refresh_interval"`
+	Service           string         `yaml:"service"`
+}
+
+// IPs struct to store IPV4 and IPV6.
+type IPs struct {
```
Comment (roidelapluie) [inline discovery/ovhcloud/ovhcloud.go:57]: Can we use netaddr.IP instead of compiling our own type?

File: discovery/ovhcloud/vps.go [hunk 34]
```diff
@@ -0,0 +1,175 @@
...
+	Vcore               int      `json:"vcore"`
+	Version             string   `json:"version"`
+	Name                string   `json:"name"`
+	Disk                int      `json:"disk"`
+	Memory              int      `json:"memory"`
+}
+
+// VPS struct from API.
+type virtualPrivateServer struct {
+	IPs                IPs      `json:"ips"`
```
Comment (roidelapluie) [inline discovery/ovhcloud/vps.go:49]: same remark, can we use a better type? like []netaddr.IP

**Comment by roidelapluie** [comment]
Can we use https://pkg.go.dev/inet.af/netaddr and remove all the IP logic in the discovery?

File: discovery/ovhcloud/ovhcloud.go (hunk 33 above)
Comment (Arnoways) [reply discovery/ovhcloud/ovhcloud.go:57]: Yes absolutely, I guess I did not find the correct methods from netip.
Thanks for your suggestion.

**Comment by roidelapluie** [comment]
Thanks!
